tasks.py

```diff
@@ -1229,6 +1231,193 @@ def ci_build_and_install_quiche(c, repo):
         break
 
+pulp_cmd_prefix = " ".join([
+    "pulp",
+    f"--base-url {os.getenv('PULP_URL', '')}",
+    f"--username {os.getenv('PULP_CI_USERNAME', '')}",
+    f"--password {os.getenv('PULP_CI_PASSWORD', '')}"
+])
+
+def run_pulp_cmd(c, cmd):
+    max_push_attempts = 3
+    attempts = 0
+    while attempts < max_push_attempts:
+        try:
+            res = c.run(f'{pulp_cmd_prefix} {cmd}')
+            if res.exited != 0:
+                raise UnexpectedExit(res)
+            return res.stdout
+        except UnexpectedExit as e:
+            attempts += 1
+            time.sleep(5)
+            print(f'Next attempt: {attempts}')
+            if attempts == max_push_attempts:
+                raise Failure(f'Error running pulp {cmd}: {e}')
+
+@task
+def validate_pulp_credentials(c):
+    # Basic pulp command that require credentials to succeed
+    repo_name = os.getenv("PULP_REPO_NAME", '')
+    cmd = f'file repository show --repository {repo_name}'
+    run_pulp_cmd(c, cmd)
+
+@task
+def pulp_upload_file_packages_by_folder(c, source):
+    repo_name = os.getenv("PULP_REPO_NAME", '')
+    for root, dirs, files in os.walk(source):
+        for path in files:
+            file = os.path.join(root, path).split('/',1)[1]
+            # file repositories have been configured with autopublish set to true
+            cmd = f'file content upload --repository {repo_name} --file {source}/{file} --relative-path {file}'
+            run_pulp_cmd(c, cmd)
+
+@task
+def pulp_create_rpm_publication(c, product, list_os_rel, list_arch):
+    rpm_distros = ["centos", "el"]
+    for os_rel in json.loads(list_os_rel):
+        if not "el-" in os_rel:
+            break
+        release = os_rel.split('-')[1]
+        for arch in json.loads(list_arch):
+            for distro in rpm_distros:
+                repo_name = f"repo-{distro}-{release}-{arch}-{product}"
+                cmd = f'rpm publication create --repository {repo_name} --checksum-type sha512'
+                run_pulp_cmd(c, cmd)
+
+@task
+def pulp_create_deb_publication(c):
+    deb_distros = ["debian", "ubuntu"]
+    for distro in deb_distros:
+        repo_name = f"repo-{distro}"
+        cmd = f'deb publication create --repository {repo_name}'
+        run_pulp_cmd(c, cmd)
+
+@task
+def pulp_upload_rpm_packages_by_folder(c, source, product):
+    rpm_distros = ["centos", "el"]
+    builds = os.listdir(source)
+
+    for build_folder in builds:
+        release = build_folder.split('.')[0].split('-')[1]
+        arch = build_folder.split('.')[1]
+        for distro in rpm_distros:
+            repo_name = f"repo-{distro}-{release}-{arch}-{product}"
+            for root, dirs, files in os.walk(f"{source}/{build_folder}"):
+                for path in files:
+                    file = os.path.join(root, path).split('/',1)[1]
+                    # Set chunk size to 500MB to avoid creating an "upload" instead of a file. Required for signing RPMs.
```

Comment:
This looks like a hack and needs to be commented as such if it is. Either by adding an `XXX` or `FIXME` in front of the comment. And also in the docs as "known issue".

Can I read more about this? Is this the official Pulp way to do this?

What I am worried about is this might change or stop working in the future and we'll be searching for a long time to find the cause.